### PR TITLE
Adding Deconvolution functionality for MXNet throwing unsupported exc…

### PR DESCRIPTION
**Pull request by euromutt**
…eptions for others.

## Channel for questions ##
Before or while filing an issue please feel free to join our [Slack channel](https://join.slack.com/t/deepjavalibrary/shared_invite/zt-ar91gjkz-qbXhr1l~LFGEIEeGBibT7w) to get in touch with development team, ask questions, find out what's cooking and more!

## Description ##
This PR adds a Deconvolution abstract class and extensions for Conv1dTranspose and Conv2dTranspose to ai.djl.nn.convolutional. NDArrayEX is revised to include a deconvolution method and the MXNet implementation MXNDArrayEX is updated with bindings to the C Libraries. Unsupported Exceptions are thrown for other engines.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- [x] Code is well-documented: 
    - For user-facing API changes, Java doc has been updated. 
    - For new examples, README.md is added to explain the what the example does.
- [x] To the my best knowledge, [examples](https://github.com/awslabs/djl/tree/master/examples) and [jupyter notebooks](https://github.com/awslabs/djl/tree/master/jupyter) are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] MXNDArrayEX, tests, (and when applicable, Java doc)
- [x] PtNDArrayEX, tests, (and when applicable, Java doc)
- [x] TfNDArrayEX, tests, (and when applicable, Java doc)
- [x] NDArrayEX, tests, (and when applicable, Java doc)

## Comments ##
- Adding transpose convolution blocks allows convolutional autoencoders and generative adversarial networks to be built in DJL.

### CODE REVIEW

**Review by frankfliu — COMMENTED** [review]
@euromutt 
I made some cosmetic changes to your PR:
1. reformatted your java code, please format your java code before commit next time:
```
./gradlew formatJava
```
2. Fixed javadoc and checkstyle violations, you can run the following command to check:
```
./gradlew check -x test
```

File: api/src/main/java/ai/djl/nn/convolutional/Deconvolution.java [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+ *
+ * <p>Current implementations of {@code Deconvolution} are {@link Conv1dTranspose} with input
+ * dimension of {@link LayoutType#WIDTH} and {@link Conv2dTranspose} with input dimension of {@link
+ * LayoutType#WIDTH} and {@link LayoutType#HEIGHT}. These implementations share the same core
+ * principal as a {@code Deconvolution} layer does, with the difference being the number of input
+ * dimension each operates on as denoted by {@code ConvXdTranspose} for {@code X} dimension(s).
+ */
+public abstract class Deconvolution extends AbstractBlock {
+
+    private static final byte VERSION = 3;
```
Comment (stu1130) [inline api/src/main/java/ai/djl/nn/convolutional/Deconvolution.java:49]: this should be version 1

File: api/src/main/java/ai/djl/nn/convolutional/Conv1dTranspose.java [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    }
+
+    /**
+     * Applies 1D deconvolution over an input signal composed of several input planes.
+     *
+     * @param input the input {@code NDArray} of shape (batchSize, inputChannel, width)
+     * @param weight filters {@code NDArray} of shape (outChannel, inputChannel/groups, width)
+     * @return the output of the conv1dtranspose operation
+     */
+    public static NDList conv1dtranspose(NDArray input, NDArray weight) {
```
Comment (stu1130) [inline api/src/main/java/ai/djl/nn/convolutional/Conv1dTranspose.java:81]:
```suggestion
    public static NDList conv1dTranspose(NDArray input, NDArray weight) {
```
same applies to the rest

**Review by stu1130 — COMMENTED** [review]
Thanks for your contribution!

Generally, it looks great!
you can add unit test on [BlockCoreTest.java](https://github.com/awslabs/djl/blob/master/integration/src/main/java/ai/djl/integration/tests/nn/BlockCoreTest.java) as Lai suggested and use the following command to run it
```
./gradlew :integration:run -Dai.djl.default_engine=MXNet
```

File: api/src/main/java/ai/djl/nn/convolutional/Deconvolution.java (hunk 1 above)
Comment (frankfliu) [inline api/src/main/java/ai/djl/nn/convolutional/Deconvolution.java:49]:
```suggestion
    private static final byte VERSION = 1;
```

File: api/src/main/java/ai/djl/nn/convolutional/Deconvolution.java [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        return new Shape[] {new Shape(shape)};
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void loadMetadata(byte version, DataInputStream is)
+            throws IOException, MalformedModelException {
+        if (version == VERSION) {
+            readInputShapes(is);
+        } else if (version != 1) {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/nn/convolutional/Deconvolution.java:157]:
```suggestion
        } else {
```

**Review by frankfliu — COMMENTED** [review]
Please add a test in BlockCoreTest.

**Comment by lanking520** [comment]
LGTM overall. The remaining part is just to add an test in the integration sections

**Comment by euromutt** [comment]
Sorry for the delay, I was in the middle of final exams. I will work the tests today or tomorrow.

Cheers,
Josh

**Comment by euromutt** [comment]
I've added tests for Conv1dTranspose and Conv2dTranspose in BlockCoreTest, but ./gradlew :integration:run -Dai.djl.default_engine=MXNet is failing (see output below). Is this a known issue?

Task :integration:run FAILED
[INFO ] - Running test ai.djl.integration.tests.model_zoo.CustomTranslatorTest ...

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':integration:run'.
> Process 'command '/usr/lib/jvm/java-11-openjdk-amd64/bin/java'' finished with non-zero exit value 139

File: api/src/main/java/ai/djl/nn/convolutional/Conv1dTranspose.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
+/*
+<<<<<<<<< Temporary merge branch 1
```
Comment (lanking520) [inline api/src/main/java/ai/djl/nn/convolutional/Conv1dTranspose.java:2]: need resolve confilict here

File: api/src/main/java/ai/djl/nn/convolutional/Conv2dTranspose.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+ * <ul>
+ *   <li>{@code data: (batch_size, channel, height, width)}
+ *   <li>{@code weight: (num_filter, channel, kernel[0], kernel[1])}
+ *   <li>{@code bias: (num_filter,)}
+ *   <li>{@code out: (batch_size, num_filter, out_height, out_width)} <br>
+<<<<<<<<< Temporary merge branch 1
+ *       {@code out_height = f(height, kernel[0], pad[0], o_pad[0], stride[0], dilate[0])} <br>
+ *       {@code out_width = f(width, kernel[1], pad[1], o_pad[1], stride[1], dilate[1])} <br>
+ *       {@code where f(x, k, p, o_p, s, d) = (x-1)*s-2*p+k+o_p}
+=========
```
Comment (lanking520) [inline api/src/main/java/ai/djl/nn/convolutional/Conv2dTranspose.java:36]: need resolve conflict here

File: build.gradle [hunk 6]
```diff
@@ -1,13 +1,16 @@
 plugins {
-    id "com.github.spotbugs" version "4.2.0" apply false
+    id "com.github.spotbugs" version "2.0.0" apply false
```
Comment (lanking520) [inline build.gradle:2]: Seemed merging causing some issue

**Review by lanking520 — CHANGES_REQUESTED** [review]
Seemed the rebase caused some issue. need to address

**Comment by lanking520** [comment]
It seemed you accidentally use an old build (DJL 0.4.0) to develop and now it conflict with the major version

**Comment by euromutt** [comment]
There was a weird merge when I added the test, I assumed it was related to the 0.8 release, but clearly that was not the case. I'm going to kill my fork and resubmit the PR.

**Pull request closed without merging**